File: inventory-service/internal/messaging/stock_reserved_contract_test.go
package messaging

import (
	"encoding/json"
	"testing"
	"time"

	"inventory-service/internal/models"
)

// Contract test: stock.reserved (RabbitMQ order.exchange), published by
// this service's Outbox (ADR-0007). Consumed by orders-service.
func TestStockReservedEventConformsToSharedSchema(t *testing.T) {
	event := models.StockReservedEvent{
		OrderID:   "order-1",
		Success:   true,
		Message:   "stock reserved",
		Timestamp: time.Now(),
	}

	body, err := json.Marshal(event)
	if err != nil {
		t.Fatalf("failed to marshal StockReservedEvent: %v", err)
	}

	var payload interface{}
	if err := json.Unmarshal(body, &payload); err != nil {
		t.Fatalf("failed to unmarshal for validation: %v", err)
	}

	schema, err := loadSchema("stock-reserved.schema.json")
	if err != nil {
		t.Fatalf("failed to load shared schema: %v", err)
	}

	if err := schema.Validate(payload); err != nil {
		t.Errorf("published StockReservedEvent violates shared schema: %v", err)
	}
}

File: orders-service/src/test/java/com/easydora/orders/consumer/StockReservedContractTest.java
package com.easydora.orders.consumer;

import com.easydora.orders.event.SchemaContractSupport;
import com.easydora.orders.service.OrderService;
import com.fasterxml.jackson.databind.JsonNode;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.ValidationMessage;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.IOException;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;

/**
 * Contract test: stock.reserved (RabbitMQ order.exchange), published by
 * inventory-service's Outbox. This consumer (InventoryEventsConsumer) has
 * no fixed DTO -- it parses raw JsonNode -- so the contract check validates
 * a schema-conformant example payload, then feeds that exact text through
 * the real production method to prove the two are actually connected.
 */
@ExtendWith(MockitoExtension.class)
class StockReservedContractTest {

    @Mock
    private OrderService orderService;

    @Test
    void consumedEventConformsToSharedSchemaAndIsHandledCorrectly() throws Exception {
        String wirePayload = "{\"orderId\":\"order-1\",\"success\":true,"
                + "\"message\":\"stock reserved\",\"timestamp\":\"2026-07-13T10:00:00Z\"}";

        JsonNode payload = SchemaContractSupport.MAPPER.readTree(wirePayload);
        JsonSchema schema = SchemaContractSupport.loadSchema("stock-reserved.schema.json");
        Set<ValidationMessage> errors = schema.validate(payload);
        assertThat(errors)
                .withFailMessage("example stock.reserved payload violates its own shared schema: %s", errors)
                .isEmpty();

        new InventoryEventsConsumer(orderService).handleStockReserved(wirePayload);

        verify(orderService).handleInventoryReserved("order-1");
    }
}
